sdk/core/azure-core/inc/azure/core/context.hpp

```diff
@@ -349,21 +146,26 @@ namespace Azure { namespace Core {
      * @return A value associated with the context found; an empty value if a specific value can't
      * be found.
      */
-    const ContextValue& operator[](const std::string& key) const
+    template <class T> const T& Get(const std::string& key) const
```

Comment:
@JeffreyRichter also suggests GetWithDefaultValue

Reply:
The impl of `GetWithDefault` can be tricky. Because get function returns a const reference, we need to consider variable lifetime issue.

```C++
const std::string& value = context.GetWithDefault("key_doesnt_exist", std::string("default_value"));
```

Reply:
There are a few things we can do here and they are related.

For the key, I'd prefer to avoid having a string field in the Context because it adds significant % overhead to the Context and I'd also prefer to not do string comparisons. Rember that we have a linked-list of contexts and so there may be several of them (which multiplies the size) and comparing each Context's string is a perf hit too. Furthermore, there is just no value to the key being a string; the key could be an integer or something else - we just need uniqueness so small & fast is worth striving for. And code that sets the key and looks up the key is written by the same person and so only this code needs to know what the key is.

So, one option is that actually make the typeid be the key. We already have this field and so WithValue can take just 1 parameter (the T) and GetValue uses typeid(T) to compare when scanning the Context linked-list. So now, the structure data type is (implicitly) the key itself: small and fast and simpler API.

If you really prefer an explicit key for some reason, then I propose a void*; just some memory address which is unique. Then the code that sets/gets the key can pass in the address to something that only it knows about (like a global Boolean variable).

If the key can't be found, then return the default. Or, if this is problematic due to the const reference, I'd change Get to TryGet and have it return a Boolean and an output T parameter.

Reply:
One more thing: If we have WithValue, then I think we should have GetValueWithDefault; not Get and not GetWithDefaultValue. I'm looking for naming consistency.

Reply:
@antkmsft 's proposal [here](https://github.com/Azure/azure-sdk-for-cpp/pull/1824#discussion_r590942383) of comparing key by pointers can work very well

Reply:
OK, Anton's proposal is fine. It looked complicated as presented and I the definition of Key was at the bottom so I never understood it. But, I've now really analyzed it and I'm happy with it too.